core/src/main/java/org/renjin/primitives/subset/Subsetting.java

```diff
@@ -339,8 +343,58 @@ public static SEXP setSingleElement(SEXP source, @ArgumentList ListVector argume
       subscripts.add(argumentList.get(i));
     }
 
+
+
+
+    if (source instanceof ListVector && isRecursiveIndexingArgument(subscripts) ) {
+
+      SEXP result = source;
+
+      if (argumentList.get(0).length() == 3) {
+        int itr = source.getIndexByName(argumentList.get(0).getElementAsSEXP(0).asString());
+        SEXP source2 = ((ListVector) source).get(itr);
+
+        int itr2 = source2.getIndexByName( argumentList.get(0).getElementAsSEXP(1).asString() );
+        SEXP source3 = ((ListVector) source2).get(itr2);
+
+        List<SEXP> sub1 = Lists.newArrayListWithCapacity(1);
+        sub1.add( argumentList.get(0).getElementAsSEXP(2) );
+        SelectionStrategy s1 = Selections.parseSingleSelection( source, sub1);
+        ListVector rep2 = s1.replaceSingleListElement( (ListVector) source3, replacement) ;
+
+
+        List<SEXP> sub2 = Lists.newArrayListWithCapacity(1);
+        sub2.add( argumentList.get(0).getElementAsSEXP(1) );
+        SelectionStrategy s2 = Selections.parseSingleSelection( source, sub2);
+        ListVector rep3 = s2.replaceSingleListElement( (ListVector) source2, rep2);
+
+        List<SEXP> sub3 = Lists.newArrayListWithCapacity(1);
+        sub3.add( argumentList.get(0).getElementAsSEXP(0) );
+        SelectionStrategy s3 = Selections.parseSingleSelection( source, sub3);
+        result = s3.replaceSingleListElement( (ListVector) source, rep3);
+      }
+
+      if (argumentList.get(0).length() == 2) {
+        int itr = source.getIndexByName(argumentList.get(0).getElementAsSEXP(0).asString());
+        SEXP source2 = ((ListVector) source).get(itr);
+
+        List<SEXP> sub2 = Lists.newArrayListWithCapacity(1);
+        sub2.add( argumentList.get(0).getElementAsSEXP(1) );
+        SelectionStrategy s2 = Selections.parseSingleSelection( source, sub2);
+        ListVector rep3 = s2.replaceSingleListElement( (ListVector) source2, replacement);
+
+        List<SEXP> sub3 = Lists.newArrayListWithCapacity(1);
+        sub3.add( argumentList.get(0).getElementAsSEXP(0) );
+        SelectionStrategy s3 = Selections.parseSingleSelection( source, sub3);
+        result = s3.replaceSingleListElement( (ListVector) source, rep3);
+
+      }
```

Comment:
If you make this a proper 

```
if(depth == 2) {

} else if(depth =3) {

} else {
    throw new EvalException("depth > 3 not implemented");
}
```
Then you will have improved upon the current version of renjin!